File: spring-webflux/src/main/java/victor/training/reactive/usecase/zipchain/Apis.java
package victor.training.reactive.usecase.zipchain;

import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

import static victor.training.reactive.Utils.sleep;

@Slf4j
class Apis {
   public Mono<A> getA(long id) {
      return Mono.fromCallable(() -> {
         log.info("getA() -- Sending expensive REST call...");
         sleep(1000);
         return new A();
      });
   }

   public Mono<B> getB(A a) {
      return Mono.fromCallable(()-> {
         log.info("getB({})", a);
         return new B();
      });
   }

   public Mono<C> getC(A a, B b) {
      return Mono.fromCallable(() -> {
         log.info("getC({},{})", a, b);
         return new C();
      });
   }


   static class A {
   }

   static class B {
   }

   static class C {
   }
}

File: spring-webflux/src/main/java/victor/training/reactive/usecase/zipchain/Apis.kt
package victor.training.reactive.usecase.zipchain

import org.slf4j.LoggerFactory
import reactor.core.publisher.Mono
import victor.training.reactive.Utils

class Apis {
    fun getA(id: Long): Mono<A> {
        return Mono.fromCallable {
            log.info("getA() -- Sending expensive REST call...")
            Utils.sleep(1000)
            A()
        }
    }

    fun getB(a: A?): Mono<B> {
        return Mono.fromCallable {
            log.info("getB({})", a)
            B()
        }
    }

    fun getC(a: A?, b: B?): Mono<C> {
        return Mono.fromCallable {
            log.info("getC({},{})", a, b)
            C()
        }
    }

    class A
    class B
    class C
    companion object {
        private val log = LoggerFactory.getLogger(Apis::class.java)
    }
}